providers/flagd/src/main/java/dev/openfeature/contrib/providers/flagd/resolver/process/storage/connector/sync/SyncStreamQueueSource.java

```diff
@@ -115,82 +111,57 @@ public void shutdown() throws InterruptedException {
 
     /** Contains blocking calls, to be used concurrently. */
     private void observeSyncStream() throws InterruptedException {
-
         log.info("Initializing sync stream observer");
 
         // outer loop for re-issuing the stream request
         // "waitForReady" on the channel, plus our retry policy slow this loop down in error conditions
         while (!shutdown.get()) {
-
             log.debug("Initializing sync stream request");
-            final GetMetadataRequest.Builder metadataRequest = GetMetadataRequest.newBuilder();
-            GetMetadataResponse metadataResponse = null;
-
-            // create a context which exists to track and cancel the stream
-            try (CancellableContext context = Context.current().withCancellation()) {
-
-                restart(); // start the stream with the context
 
-                // TODO: remove the metadata call entirely after https://github.com/open-feature/flagd/issues/1584
-                if (!syncMetadataDisabled) {
-                    try {
-                        FlagSyncServiceBlockingStub localStub = blockingStub;
-
-                        if (deadline > 0) {
-                            localStub = localStub.withDeadlineAfter(deadline, TimeUnit.MILLISECONDS);
-                        }
-
-                        metadataResponse = localStub.getMetadata(metadataRequest.build());
-                    } catch (Exception metaEx) {
-                        // cancel the stream if the getMetadata fails
-                        // we can keep this log quiet since the stream is cancelled/restarted with this exception
-                        log.debug("Metadata exception: {}, cancelling stream", metaEx.getMessage(), metaEx);
-                        context.cancel(metaEx);
+            try (Context.CancellableContext context = Context.current().withCancellation()) {
+                SyncStreamObserver observer = new SyncStreamObserver(outgoingQueue, context);
+                Context previous = context.attach();
+                try {
+                    if (!syncMetadataDisabled) {
+                        observer.metadata = getMetadata(context);
```

Comment:
```suggestion
                        observer.getMetadata(context);
```
This goes together with my other comment about moving the `syncFlags` method into the `SyncStreamObserver`. I think it would be cleaner to move the functionality into the object.

Reply:
See other comment